File: arxiv_downloader.js
import fs from 'fs';
import path from 'path';

/**
 * Node.js arXiv Downloader (No API Key Required)
 * Uses arXiv's public API to query papers and save metadata / PDFs
 */
async function fetchArxivPapers(query = "Retrieval Augmented Generation", maxResults = 10, downloadPdfs = true) {
  console.log(`🔍 Searching arXiv for: "${query}" (Max results: ${maxResults})...\n`);

  const apiUrl = `http://export.arxiv.org/api/query?search_query=all:${encodeURIComponent(query)}&start=0&max_results=${maxResults}&sortBy=relevance&sortOrder=descending`;

  try {
    const response = await fetch(apiUrl);
    const xmlText = await response.text();

    // Simple regex parser for Atom XML entries
    const entries = xmlText.split('<entry>').slice(1);
    const papers = [];

    const pdfDir = path.join(process.cwd(), 'data', 'papers');
    if (downloadPdfs && !fs.existsSync(pdfDir)) {
      fs.mkdirSync(pdfDir, { recursive: true });
    }

    for (let i = 0; i < entries.length; i++) {
      const entry = entries[i];
      const title = (entry.match(/<title>([\s\S]*?)<\/title>/) || [])[1]?.replace(/\s+/g, ' ').trim() || 'Untitled';
      const idUrl = (entry.match(/<id>([\s\S]*?)<\/id>/) || [])[1]?.trim() || '';
      const paperId = idUrl.split('/abs/')[1] || `paper_${i + 1}`;
      const published = (entry.match(/<published>([\s\S]*?)<\/published>/) || [])[1]?.substring(0, 10) || '';
      const summary = (entry.match(/<summary>([\s\S]*?)<\/summary>/) || [])[1]?.replace(/\s+/g, ' ').trim() || '';
      
      const authorMatches = [...entry.matchAll(/<name>([\s\S]*?)<\/name>/g)];
      const authors = authorMatches.map(m => m[1].trim());

      const pdfUrl = `https://arxiv.org/pdf/${paperId}.pdf`;

      const paperData = { paperId, title, authors, published, summary, pdfUrl };
      papers.push(paperData);

      console.log(`[${i + 1}] Title: ${title}`);
      console.log(`    Authors: ${authors.slice(0, 3).join(', ')}${authors.length > 3 ? ' et al.' : ''}`);
      console.log(`    Published: ${published}`);
      console.log(`    PDF URL: ${pdfUrl}`);

      if (downloadPdfs) {
        const safeTitle = title.replace(/[^a-zA-Z0-9 _-]/g, '').trim().substring(0, 40);
        const filename = `${paperId.replace('/', '_')}_${safeTitle}.pdf`;
        const filePath = path.join(pdfDir, filename);

        if (!fs.existsSync(filePath)) {
          console.log(`    📥 Downloading PDF to: ${filePath}`);
          try {
            const pdfRes = await fetch(pdfUrl);
            if (pdfRes.ok) {
              const arrayBuffer = await pdfRes.arrayBuffer();
              fs.writeFileSync(filePath, Buffer.from(arrayBuffer));
            } else {
              console.log(`    ⚠️ Could not download PDF (Status: ${pdfRes.status})`);
            }
          } catch (err) {
            console.log(`    ⚠️ Error downloading PDF: ${err.message}`);
          }
        } else {
          console.log(`    ✅ PDF already downloaded.`);
        }
      }
      console.log('-'.repeat(75));
    }

    const dataDir = path.join(process.cwd(), 'data');
    if (!fs.existsSync(dataDir)) fs.mkdirSync(dataDir, { recursive: true });

    const metadataPath = path.join(dataDir, 'papers_metadata.json');
    fs.writeFileSync(metadataPath, JSON.stringify(papers, null, 2));

    console.log(`\n✨ Successfully fetched ${papers.length} papers.`);
    console.log(`📁 Metadata saved to: ${metadataPath}`);
    if (downloadPdfs) console.log(`📁 PDFs saved to: ${pdfDir}`);

  } catch (error) {
    console.error('❌ Failed to fetch arXiv papers:', error);
  }
}

fetchArxivPapers("Retrieval Augmented Generation", 10, true);

File: arxiv_downloader.py
import arxiv
import os
import json
import ssl
from datetime import datetime

# Bypass SSL certificate verification issues (common in local Python setups)
try:
    ssl._create_default_https_context = ssl._create_unverified_context
except AttributeError:
    pass

def fetch_arxiv_papers(query="Retrieval Augmented Generation", max_results=10, download_pdfs=False):
    """
    Fetches paper metadata from arXiv based on a search query.
    Optionally downloads PDFs into a local folder.
    """
    print(f"🔍 Searching arXiv for: '{query}' (Max results: {max_results})...\n")
    
    client = arxiv.Client()
    search = arxiv.Search(
        query=query,
        max_results=max_results,
        sort_by=arxiv.SortCriterion.Relevance,
    )

    papers_metadata = []
    
    # Create directory for PDFs if download is requested
    pdf_dir = os.path.join(".", "data", "papers")
    if download_pdfs:
        os.makedirs(pdf_dir, exist_ok=True)

    for i, result in enumerate(client.results(search), 1):
        metadata = {
            "paper_id": result.entry_id.split("/")[-1],
            "title": result.title,
            "authors": [author.name for author in result.authors],
            "published": result.published.strftime("%Y-%m-%d"),
            "summary": result.summary,
            "pdf_url": result.pdf_url,
            "categories": result.categories
        }
        papers_metadata.append(metadata)

        print(f"[{i}] Title: {metadata['title']}")
        print(f"    Authors: {', '.join(metadata['authors'][:3])}{' et al.' if len(metadata['authors']) > 3 else ''}")
        print(f"    Published: {metadata['published']}")
        print(f"    PDF URL: {metadata['pdf_url']}")
        
        if download_pdfs:
            # Clean title for filename
            safe_title = "".join(c if c.isalnum() or c in (' ', '_', '-') else '' for c in metadata['title']).strip()
            filename = f"{metadata['paper_id']}_{safe_title[:40]}.pdf"
            filepath = os.path.join(pdf_dir, filename)
            
            if not os.path.exists(filepath):
                print(f"    📥 Downloading PDF to: {filepath}")
                from urllib.request import urlretrieve
                urlretrieve(metadata['pdf_url'], filepath)
            else:
                print(f"    ✅ PDF already downloaded.")
        
        print("-" * 75)

    # Save metadata summary as JSON for future chunking / embedding pipeline
    os.makedirs(os.path.join(".", "data"), exist_ok=True)
    json_path = os.path.join(".", "data", "papers_metadata.json")
    with open(json_path, "w", encoding="utf-8") as f:
        json.dump(papers_metadata, f, indent=2)

    print(f"\n✨ Successfully fetched metadata for {len(papers_metadata)} papers.")
    print(f"📁 Metadata saved to: {json_path}")
    if download_pdfs:
        print(f"📁 PDFs downloaded to: {pdf_dir}")

if __name__ == "__main__":
    # Fetch 50 papers related to RAG and save metadata + PDFs
    fetch_arxiv_papers(query="Retrieval Augmented Generation", max_results=50, download_pdfs=True)
